### PR TITLE
MM force field refactor and OPLS-AA implementation

### PR DESCRIPTION
**Pull request by DRosen285**
Intial commit to incorporate the OPLS-AA force field in jax-MD

### CODE REVIEW

**Comment by beitom** [comment]
Great to hear that the project is alive @ekindogus! I just saw many open PRs and not much recent activity. Adding classical forcefields to jax-md could be a good way to get it back going as we already have this open PR and the related on https://github.com/jax-md/jax-md/pull/319 for AMBER. Are `jax-md` devs interested in this addition? I could also see a separate standalone library e.g. `jax-ff` that focuses on just on ffs (AMBER, CHARM, oplsaa) and related tools like PME.

**Comment by beitom** [comment]
Same with me, I think this a cool project and would be great to get it to a state where we could run some utility scale MD workflows. Reproducing some of the benchmarks in [MACE-OFF](https://arxiv.org/pdf/2312.15211) (see sections C and D) could be a good goal but ambitious longer terms goal. How do you guys want to continue this discussion?

**Comment by beitom** [comment]
@abhijeetgangan thanks! it's tom@beit.tech

**Comment by alecbetancourt** [comment]
Feel free to reach out to me at my personal email (alecbetancourt@gmail.com)!

**Comment by abhijeetgangan** [comment]
Great, hopefully everyone got the email.

**Comment by abhijeetgangan** [comment]
I think that looks good. Few suggestions would be that to not call it forcefield as it is quite generic. I was thinking it could be particularly for MM style forcefield. Also, we can probably do better than having multiple `energy.py` modules?

**Comment by abhijeetgangan** [comment]
I think `jax_md.mm_forcefields` should be good. Regarding David's comment on handling topology files I do not have a strong opinion. I agree that we should keep it modular as possible.

We could start structuring the current PR and see where it goes?

Side note - I am also thinking of having some IO connecting pymatgen and ase so we could read / write different formats.

**Comment by beitom** [comment]
> Regarding the file parsers: I'm mostly familiar with gromacs and lammps. So I would say .top/.itp files from gromacs and lammps.data fiels. For AMBER, CHARMM or NAMD I'm not so familiar with the topology file system. I already have written a lammps parser (see: https://github.com/DRosen285/OPLSAA_for_jax-MD/blob/main/extract_params_oplsaa.py). This parsers is actually quite generic, because it just extracts what is inside the lammps.data file. So it is not limited to OPLS.
> 
> Let's start structuring the current PR :)

@DRosen285 do you have any sample data we could use here? I was thinking of trying to run a simple experiment for torsion profiling which could also be a good notebook demo to add once this feature is complete.

I spent a bit of time yesterday working on the following structure, please let me know your feedback:
```
mm_forcefields/
├── base.py
├── neighbor.py
├── nonbonded
│   ├── electrostatics.py
│   └── __init__.py
├── oplsaa
│   ├── energy.py
│   ├── __init__.py
│   ├── io.py
│   ├── params.py
│   └── topology.py
└── reaxff
    ├── ... # (can move here in a future MR)
```

Current structure/example with oplsaa.energy
```python
(function) def energy(
    topology: Topology,
    params: Parameters,
    box: Array,
    coulomb: CoulombHandler,
    nb_options: NonbondedOptions = NonbondedOptions()
) -> Tuple[(...) -> Unknown, (...) -> Unknown, (...) -> Unknown]
Create OPLSAA energy function.

Args:
    topology: Molecular topology.
    params: Force field parameters.
    box: Simulation box (scalar or array).
    coulomb: Coulomb energy handler.
    nb_options: Nonbonded interaction options.

Returns:
    energy_fn: Function that computes energy given positions and neighbor list.
        Signature: energy_fn(positions, nlist) -> energy_dict
    neighbor_fn: Function to build/update neighbor lists.
    displacement_fn: Displacement function for the given box.

Example:

>>> from jax_md.mm import oplsaa
>>> from jax_md.mm.nonbonded.electrostatics import PMECoulomb
>>>
>>> top = oplsaa.Topology(...)
>>> params = oplsaa.Parameters(...)
>>> coulomb = PMECoulomb(grid_size=32, alpha=0.3, r_cut=12.0)
>>> nb_opts = oplsaa.NonbondedOptions(r_cut=12.0, dr_threshold=0.5)
>>>
>>> energy_fn, neighbor_fn, disp_fn = oplsaa.energy(
...     top, params, box, coulomb, nb_opts
... )
>>>
>>> # Initialize neighbor list
>>> nlist = neighbor_fn.allocate(positions)
>>>
>>> # Compute energy
>>> E = energy_fn(positions, nlist)
```

**Comment by DRosen285** [comment]
@beitom Examples can be found here; https://github.com/DRosen285/OPLSAA_for_jax-MD/tree/main/examples 
As the names suggest "single molecule" contains data of a single 10 atom molecule, and in "bulk system" you find multiple system sizes for that molecule. The largest one "EC_bulk.data" has 7500 atoms (750 molecules).

**Comment by beitom** [comment]
> @beitom Examples can be found here; https://github.com/DRosen285/OPLSAA_for_jax-MD/tree/main/examples As the names suggest "single molecule" contains data of a single 10 atom molecule, and in "bulk system" you find multiple system sizes for that molecule. The largest one "EC_bulk.data" has 7500 atoms (750 molecules).

Great, do you think it would be worthwhile to generate some reference data for this system using lamps or open mm so we have a basic benchmark for the jax implementation?

**Comment by beitom** [comment]
> ```python
> Topology
> ```



> @beitom That looks good! Maybe `neighbors.py` can be outside since it is more general? Would be nice to follow the same structure for the unit tests.
> 
> Also, the molecule testing is fine but for bulk we need to be careful about the minimum image convention.

@abhijeetgangan I haven't completed it yet but I think neighbors.py will somewhat specific content for mmff into easily generalizable e.g. concepts like 1-2, 1-3, 1-4 exclusions.

By the way do you have a preferred way for me to contribute to this PR (through another PR or directly onto this one if im given permissions)

File: jax_md/mm_forcefields/README.md
Comment (beitom) [inline jax_md/mm_forcefields/README.md]: Please start review with this README

**Comment by abhijeetgangan** [comment]
@beitom This looks great! I will take a detailed review later once you think it is finished. Would be nice to also include some tests. I think @DRosen285 can help with the oplsaa tests.

File: jax_md/mm_forcefields/neighbor.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+      valid_14_pairs[:, 0], valid_14_pairs[:, 1]
+    ].set(True)
+    pair_14_mask = pair_14_mask.at[
+      valid_14_pairs[:, 1], valid_14_pairs[:, 0]
+    ].set(True)
+
+  return pair_14_mask
+
+
+def safe_norm(
```
Comment (beitom) [inline jax_md/mm_forcefields/neighbor.py]: TODO: decode where to move these utils to @abhijeetgangan

**Comment by beitom** [comment]
> @beitom This looks great! I will take a detailed review later once you think it is finished. Would be nice to also include some tests. I think @DRosen285 can help with the oplsaa tests.

Thanks @abhijeetgangan! From my perspective this is ready for review, and I started with a few comments. One of the bigger decisions to make it what kinds of systems we want to support. For now I just made a quick and simple CHARMM parser to we could run some basic examples using some benchmark data I found (https://github.com/RowleyGroup/torsionbenchmark/tree/master/opls). You can see the example in the [notebook](https://github.com/DRosen285/jax-md/blob/340d4f95e27169d48a1f024bb4a118fb8fa5a7a6/notebooks/oplssa_torsion.ipynb). Im sure we could integrate @DRosen285  LAMPS parser as well. However tho I think the most modern approach would be, ditch these custom parsers and to provide some integration with [openff](https://github.com/openforcefield/openff-toolkit). I sketched out a basic pipeline [here](https://www.figma.com/board/EZjQPwK5toe3W0LFIYRroG/OpenFF_JAXMD_Workflow?node-id=0-1&t=o7p08EtqnP7Ilwgy-1). @DRosen285 I'm interested in your opinion on this.

File: tests/oplsaa_test.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    self.assertIsNotNone(self.topology)
+    self.assertIsNotNone(self.parameters)
+
+    self.assertEqual(self.positions.shape[1], 3)
+    self.assertEqual(self.topology.n_atoms, self.positions.shape[0])
+    self.assertGreater(self.topology.n_atoms, 0)
+
+  def test_energy_computation(self):
+    """Test that energy can be computed."""
+    E = self.energy_fn(self.positions, self.nlist)
```
[the remark this answers is not in the document]
Comment (DRosen285) [reply tests/oplsaa_test.py]: @abhijeetgangan I did add a print statement for the individual contributions to the test. Do you want me to also include the hard-codeed lammps energies?

[the remark this answers is not in the document]
Comment (abhijeetgangan) [reply tests/oplsaa_test.py]: Hard coded values are fine. Something like - https://github.com/jax-md/jax-md/blob/6bd17d29ce5f9fe35a5582a42a9973b1ecd0859f/tests/energy_test.py#L956

File: tests/oplsaa_test.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  def test_gradient_computation(self):
+    """Test that gradients can be computed."""
+
+    def total_energy_fn(pos, nlist):
+      E = self.energy_fn(pos, nlist)
+      return E['total']
+
+    grad_fn = grad(total_energy_fn, argnums=0)
+    gradients = grad_fn(self.positions, self.nlist)
```
[the remark this answers is not in the document]
Comment (DRosen285) [reply tests/oplsaa_test.py]: Added now a test for stress computation. This only works for CutoffCoulomb. It seems that the reciprocal-space grid for PME methods makes the energy discontinuous w.r.t box.

File: tests/oplsaa_test.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      "Angle": 0.21927227,
+      "Torsion": 0.0,
+      "Improper": 3.1140185e-05,
+      "vdwl":  11.911979,
+      "Coulomb": -1.0519856,
+      "Total": 12.480843
+     }
+
+    # Print energy terms
+    print("\nEnergy terms:")
```
Comment (abhijeetgangan) [inline tests/oplsaa_test.py]: The assertions are good enough. Printing is not required as such.

**Comment by abhijeetgangan** [comment]
@beitom @DRosen285 I think this is good to merge after the tests. Let me know if you guys agree and we can go ahead. Thanks both!

**Comment by abhijeetgangan** [comment]
Sounds good. Thanks a lot!
